File: Data Structures & Algorithms/search-for-word/submission-10.js
class Solution {
    /**
     * @param {character[][]} board
     * @param {string} word
     * @return {boolean}
     */
    exist(board, word) { 
        function dfs(i, r, c) {
            if (i === word.length) return true 

            if (r < 0 ||   
                c < 0 || 
                r >= board.length || 
                c >= board[0].length || 
                board[r][c] !== word[i] || 
                board[r][c] === '.') {
                    return false
                } 
            
            const temp = board[r][c]
            board[r][c] = '.' 

            const res = dfs(i + 1, r + 1, c) || 
                        dfs(i + 1, r - 1, c) || 
                        dfs(i + 1, r, c + 1) || 
                        dfs(i + 1, r, c - 1)  

            board[r][c] = temp 

            return res
        } 

        for (let r = 0; r < board.length; r ++) {
            for (let c = 0; c < board[0].length; c ++) {
                const cur = board[r][c] 
                if (cur === word[0]) {
                    if (dfs(0, r, c)) return true
                }
            }
        } 

        return false
    }
}

File: Data Structures & Algorithms/permutations/submission-10.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        added = set()  
        res = []

        def dfs(cur): 
            if len(cur) == len(nums): 
                res.append(cur.copy()) 
                return 

            for i in range(len(nums)): 
                curNum = nums[i] 
                if curNum in added: 
                    continue 
                added.add(curNum) 
                cur.append(curNum) 
                dfs(cur) 
                added.discard(curNum) 
                cur.pop() 

        dfs([]) 
        return res